<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>R Basics Continued</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/computer/
-->
<meta name="google-site-verification" content="t-Wgs-RXt8xi94RSoFamxoD2P2h-Hh7M_t_iOuxFZUo" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Computational Foundations</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Bash Shell
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="bash-01-introduction.html">Introducing the Shell</a>
    </li>
    <li>
      <a href="bash-02-the-filesystem.html">Navigating Files and Directories</a>
    </li>
    <li>
      <a href="bash-03-working-with-files.html">Working with Files</a>
    </li>
    <li>
      <a href="bash-04-redirection.html">Redirection</a>
    </li>
    <li>
      <a href="bash-05-writing-scripts.html">Writing Scripts and Working with Data</a>
    </li>
    <li>
      <a href="bash-06-organization.html">Project Organization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    R and RStudio
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="r-01-introduction.html">Introducing R and RStudio IDE</a>
    </li>
    <li>
      <a href="r-02-r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-03-basics-continued.html">R Basics Continued</a>
    </li>
    <li>
      <a href="r-04-bioconductor.html">Using Packages from Bioconductor</a>
    </li>
    <li>
      <a href="r-05-dplyr.html">Data Wrangling and Analyses with Tidyverse</a>
    </li>
    <li>
      <a href="r-06-data-visualization.html">Data Visualizations with ggplot2</a>
    </li>
    <li>
      <a href="r-07-r-help.html">Getting Help with R</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">R Basics Continued</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#objectives" id="toc-objectives">Objectives</a></li>
<li><a href="#factors" id="toc-factors">Factors</a></li>
<li><a href="#data-frames" id="toc-data-frames">Data Frames</a></li>
<li><a href="#working-with-spreadsheets"
id="toc-working-with-spreadsheets">Working with spreadsheets</a>
<ul>
<li><a href="#importing-tabular-data-into-r"
id="toc-importing-tabular-data-into-r">Importing tabular data into
R</a></li>
<li><a href="#summarizing-subsetting-and-reordering-a-data-frame"
id="toc-summarizing-subsetting-and-reordering-a-data-frame">Summarizing,
subsetting, and reordering a data frame</a>
<ul>
<li><a href="#summarizing" id="toc-summarizing">Summarizing</a></li>
<li><a href="#subsetting" id="toc-subsetting">Subsetting</a></li>
<li><a href="#reordering" id="toc-reordering">Reordering</a></li>
</ul></li>
<li><a href="#coercing-data" id="toc-coercing-data">Coercing
data</a></li>
<li><a href="#saving-your-data-frame-to-a-file"
id="toc-saving-your-data-frame-to-a-file">Saving your data frame to a
file</a></li>
</ul></li>
<li><a href="#data-tips" id="toc-data-tips">Data tips</a></li>
</ul>
</div>

<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

</style>
<div id="objectives" class="section level1">
<h1>Objectives</h1>
<p>In this lesson we will introduce two more data types: factors and
data frames. Factors are the data type R uses to store categorical data.
Data frames are probably the most commonly used data type in R, and is
most similar to Excel spreadsheets. We will learn how to import data
into R as a data frame, and in subsequent lessons how to manipulate,
modify, summarize, and plot data contained in data frames.</p>
</div>
<div id="factors" class="section level1">
<h1>Factors</h1>
<p>Factors are how R stores categorical information, like the
continents, among our examples. It is like a character vector that can
only have a finite number of values. To make a factor, we can pass an
atomic vector into the <code>factor()</code> function. In the
background, R recodes the data as integers and stores the results in an
integer vector. It also adds a <code>levels</code> attribute to the
integer vector enumerating the set of labels for displaying the factor
values. The easiest way to see this is with an example:</p>
<pre class="r"><code>continents_factor = factor(continents)
continents_factor</code></pre>
<pre><code>[1] Asia          Africa        South America North America
Levels: Africa Asia North America South America</code></pre>
<pre class="r"><code>typeof(continents_factor)</code></pre>
<pre><code>[1] &quot;integer&quot;</code></pre>
<pre class="r"><code>class(continents_factor)</code></pre>
<pre><code>[1] &quot;factor&quot;</code></pre>
<pre class="r"><code>as.integer(continents_factor)</code></pre>
<pre><code>[1] 2 1 4 3</code></pre>
<p>We observe a few things with this string of commands:</p>
<ol style="list-style-type: decimal">
<li>It looks like the type of <code>continents_factor</code> is an
integer vector.</li>
<li>But there is a class of vectors, called factors. R will often take
into account the class of an object to decide how a function will act on
it.</li>
<li>By default the <code>levels</code> of a factor are the unique
elements of the character vector in alphabetical order.</li>
</ol>
<p>You can specify an order by explicitly defining the levels as in:</p>
<pre class="r"><code>continents_factor = factor(continents, levels = c(&#39;South America&#39;, &#39;Asia&#39;, &#39;Africa&#39;, &#39;North America&#39;))
continents_factor</code></pre>
<pre><code>[1] Asia          Africa        South America North America
Levels: South America Asia Africa North America</code></pre>
<pre class="r"><code>as.integer(continents_factor)</code></pre>
<pre><code>[1] 2 3 1 4</code></pre>
<p>Notice that the integer representation of the vector changed order,
because the levels changed order. This will be useful later in the
RNA-seq Demystified lessons when we discuss testing for differential
expression and the assumptions DESeq2 makes about what factor level is
the “reference”.</p>
<blockquote>
<h2
id="tip-treating-objects-as-categories-without-changing-their-mode">Tip:
Treating objects as categories without changing their mode</h2>
<p>You don’t have to make an object a factor to get the benefits of
treating an object as a factor. See what happens when you use the
<code>as.factor()</code> function on <code>continents</code>. To
generate a tally, you can sometimes also use the <code>table()</code>
function; though sometimes you may need to combine both
(i.e. <code>table(as.factor(continents))</code>)</p>
</blockquote>
</div>
<div id="data-frames" class="section level1">
<h1>Data Frames</h1>
<p>Data frames group vectors together into a two-dimensional table,
where each vector becomes a column of the table. The data types within
each column must be the same (they are vectors after all), <strong>but
the columns can be of different data types</strong>. Let’s construct our
first data frame using some vectors we have laying around. Here we will
include column names from the start:</p>
<pre class="r"><code>countries_df = data.frame(country = countries, continent = continents, population = populations)
countries_df</code></pre>
<pre><code>   country     continent population
1 Thailand          Asia   69950850
2    Ghana        Africa    2108328
3 Suriname South America     575990
4   Canada North America   38436447</code></pre>
<pre class="r"><code>dim(countries_df)</code></pre>
<pre><code>[1] 4 3</code></pre>
<p>Since data frames are two-dimensional, we have to specify both the
row and column to extract specific entries, as in:</p>
<pre class="r"><code># Return the element in the second row, third column
countries_df[2, 3]</code></pre>
<pre><code>[1] 2108328</code></pre>
<p>Since the columns of the data frame are named, we can also use those
instead:</p>
<pre class="r"><code># Return the element in the Ghana row of the population column
countries_df[2, &#39;population&#39;]</code></pre>
<pre><code>[1] 2108328</code></pre>
<p>To access an entire row or column, we specify just the row or column
index, as with:</p>
<pre class="r"><code># Access the third row by index
countries_df[3, ]</code></pre>
<pre><code>   country     continent population
3 Suriname South America     575990</code></pre>
<p>And it is the same with columns:</p>
<pre class="r"><code># Access the second column by index
countries_df[, 2]</code></pre>
<pre><code>[1] &quot;Asia&quot;          &quot;Africa&quot;        &quot;South America&quot; &quot;North America&quot;</code></pre>
<pre class="r"><code># By name
countries_df[, &#39;continent&#39;]</code></pre>
<pre><code>[1] &quot;Asia&quot;          &quot;Africa&quot;        &quot;South America&quot; &quot;North America&quot;</code></pre>
<p>Finally, data frames have another way to access columns using
“dollar-sign notation”:</p>
<pre class="r"><code>countries_df$continent</code></pre>
<pre><code>[1] &quot;Asia&quot;          &quot;Africa&quot;        &quot;South America&quot; &quot;North America&quot;</code></pre>
</div>
<div id="working-with-spreadsheets" class="section level1">
<h1>Working with spreadsheets</h1>
<p>A substantial amount of data is tabular, that is data arranged in
rows and columns - also known as spreadsheets. We could write a whole
lesson on how to work with spreadsheets effectively (<a
href="https://datacarpentry.org/organization-genomics/">actually, Data
Carpentry has</a>). For our purposes, we want to remind you of a few
principles before we work with our first set of example data:</p>
<p><strong>1. Keep raw data separate from analyzed data</strong></p>
<p>This is principle number one because if you can’t tell which files
are the original raw data, you risk making some serious mistakes
(e.g. drawing conclusion from data which have been manipulated in some
unknown way).</p>
<p><strong>2. Keep spreadsheet data Tidy</strong></p>
<p>The simplest principle of <strong>Tidy data</strong> is that we have
one row in our spreadsheet for each observation or sample, and one
column for every variable that we measure or report on. As simple as
this sounds, it’s very easily violated. Most data scientists agree that
significant amounts of their time is spent tidying data for analysis.
Read more about data organization in <a
href="https://datacarpentry.org/organization-genomics/">this lesson</a>
and in <a href="https://www.jstatsoft.org/article/view/v059i10">this
paper</a>.</p>
<p><strong>3. Trust but verify</strong></p>
<p>Finally, you don’t need to be paranoid about data, but you should
have a plan for how you will prepare it for analysis. You probably
already have a lot of intuition, expectations, assumptions about your
data - the range of values you expect, how many values should have been
recorded, etc. Of course, as the data get larger our human ability to
keep track will start to fail (it can fail for small data sets too). R
can help you to examine your data so that you can have greater
confidence in your analysis, and its reproducibility.</p>
<div id="importing-tabular-data-into-r" class="section level2">
<h2>Importing tabular data into R</h2>
<p>There are several ways to import data into R. For our purpose here,
we will focus on using the tools every R installation comes with (so
called “base” R) to import a comma-delimited file containing the results
of our variant calling workflow. We will need to load the sheet using a
function called <code>read.csv()</code>.</p>
<blockquote>
<h2
id="exercise-review-the-arguments-of-the-read.csv-function">Exercise:
Review the arguments of the <code>read.csv()</code> function</h2>
<p><strong>Before using the <code>read.csv()</code> function, use R’s
help feature to answer the following questions</strong>.</p>
<p><em>Hint</em>: Entering ‘?’ before the function name and then running
that line will bring up the help documentation. Also, when reading this
particular help be careful to pay attention to the ‘read.csv’ expression
under the ‘Usage’ heading. Other answers will be in the ‘Arguments’
heading.</p>
<ol style="list-style-type: decimal">
<li>What is the default parameter for ‘header’ in the
<code>read.csv()</code> function?</li>
<li>What argument would you have to change to read a file that was
delimited by semicolons (;) rather than commas?</li>
<li>What argument would you have to change to read file in which numbers
used commas for decimal separation (i.e. 1,00)?</li>
<li>What argument would you have to change to read in only the first
10,000 rows of a very large file?</li>
</ol>
</blockquote>
<details>
<summary>
Solution
</summary>
<ol style="list-style-type: decimal">
<li>The <code>read.csv()</code> function has the argument ‘header’ set
to TRUE by default, this means the function always assumes the first row
is header information, (i.e. column names)</li>
<li>The <code>read.csv()</code> function has the argument ‘sep’ set to
“,”. This means the function assumes commas are used as delimiters, as
you would expect. Changing this parameter (e.g. <code>sep=";"</code>)
would now interpret semicolons as delimiters.</li>
<li>Although it is not listed in the <code>read.csv()</code> usage,
<code>read.csv()</code> is a “version” of the function
<code>read.table()</code> and accepts all its arguments. If you set
<code>dec=","</code> you could change the decimal operator. We’d
probably assume the delimiter is some other character.</li>
<li>You can set <code>nrow</code> to a numeric value
(e.g. <code>nrow=10000</code>) to choose how many rows of a file you
read in. This may be useful for very large files where not all the data
is needed to test some data cleaning steps you are applying.</li>
</ol>
Hopefully, this exercise gets you thinking about using the provided help
documentation in R. There are many arguments that exist, but which we
wont have time to cover. Look here to get familiar with functions you
use frequently, you may be surprised at what you find they can do.
</details>
<p><br></p>
<p>Now, let’s read in the file <code>data/gapminder_data.csv</code> and
call this data <code>gapminder</code>. The first argument to pass to our
<code>read.csv()</code> function is the file path for our data. The file
path must be in quotes and now is a good time to remember to use tab
autocompletion. <strong>If you use tab autocompletion you avoid typos
and errors in file paths.</strong> Use it!</p>
<pre class="r"><code>## read in a CSV file and save it as &#39;gapminder&#39;
gapminder &lt;- read.csv(&quot;data/gapminder_data.csv&quot;)</code></pre>
<p>One of the first things you should notice is that in the Environment
window, you have the <code>gapminder</code> object, listed as 1704 obs.
(observations/rows) of 6 variables (columns). Double-clicking on the
name of the object will open a view of the data in a new tab.</p>
<p><img src="images/rstudio_dataframeview.png" alt="rstudio data frame view" style="width: 1000px;"/></p>
<blockquote>
<h2 id="tip-changes-in-r-dont-overwrite-the-original-file">Tip: Changes
in R don’t overwrite the original file</h2>
<p>When you work with data in R, you are not changing the original file
you loaded that data from. This is different than (for example) working
with a spreadsheet program where changing the value of the cell leaves
you one “save”-click away from overwriting the original file. You have
to purposely use a writing function (e.g. <code>write.csv()</code>) to
save data loaded into R. In that case, be sure to save the manipulated
data into a new file. More on this later in the lesson.</p>
</blockquote>
</div>
<div id="summarizing-subsetting-and-reordering-a-data-frame"
class="section level2">
<h2>Summarizing, subsetting, and reordering a data frame</h2>
<p>A <strong>data frame is the standard way in R to store tabular
data</strong>. A data frame is a collection of vectors of the same
length. Let’s use the <code>str()</code> (structure) function to look a
little more closely at how data frames work:</p>
<pre class="r"><code>## get the structure of a data frame
str(gapminder)</code></pre>
<pre><code>&#39;data.frame&#39;:   1704 obs. of  6 variables:
 $ country  : chr  &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; ...
 $ year     : int  1952 1957 1962 1967 1972 1977 1982 1987 1992 1997 ...
 $ pop      : num  8425333 9240934 10267083 11537966 13079460 ...
 $ continent: chr  &quot;Asia&quot; &quot;Asia&quot; &quot;Asia&quot; &quot;Asia&quot; ...
 $ lifeExp  : num  28.8 30.3 32 34 36.1 ...
 $ gdpPercap: num  779 821 853 836 740 ...</code></pre>
<p>Some things to notice:</p>
<ul>
<li>the object type <code>data.frame</code> is displayed in the first
row along with its dimensions, in this case 1704 observations (rows) and
6 variables (columns)</li>
<li>Each variable (column) has a name (e.g. <code>country</code>). This
is followed by the object mode (e.g. chr, int, etc.). Notice that before
each variable name there is a <code>$</code> - which is a hint as to how
one can access these columns, as we saw when we first introduced data
frames.</li>
</ul>
<div id="summarizing" class="section level3">
<h3>Summarizing</h3>
<p>We can get a birds eye view of a data frame by using the
<code>summary()</code> function. Depending on the type of data in the
columns, <code>summary()</code> will do particular things:</p>
<pre class="r"><code>## get summary statistics on a data frame
summary(gapminder)</code></pre>
<pre><code>   country               year           pop             continent            lifeExp        gdpPercap       
 Length:1704        Min.   :1952   Min.   :6.001e+04   Length:1704        Min.   :23.60   Min.   :   241.2  
 Class :character   1st Qu.:1966   1st Qu.:2.794e+06   Class :character   1st Qu.:48.20   1st Qu.:  1202.1  
 Mode  :character   Median :1980   Median :7.024e+06   Mode  :character   Median :60.71   Median :  3531.8  
                    Mean   :1980   Mean   :2.960e+07                      Mean   :59.47   Mean   :  7215.3  
                    3rd Qu.:1993   3rd Qu.:1.959e+07                      3rd Qu.:70.85   3rd Qu.:  9325.5  
                    Max.   :2007   Max.   :1.319e+09                      Max.   :82.60   Max.   :113523.1  </code></pre>
<p>Our data frame has 6 variables, so we get 6 fields that summarize the
data. The <code>year</code>, <code>pop</code>, <code>lifeExp</code> and
<code>gdpPercap</code> variables are numerical data and so you get
summary statistics on the min and max values for these columns, as well
as mean, median, and interquartile ranges. The other variables,
<code>country</code> and <code>continent</code>, are treated as
characters data (more on this in a bit).</p>
</div>
<div id="subsetting" class="section level3">
<h3>Subsetting</h3>
<p>We saw how to subset a data frame in the previous section, but in all
those examples we printed values to the screen. You can create a new
data frame object by assigning them to a new object name:</p>
<pre class="r"><code># create a new data frame containing only observations from India
india_subset &lt;- gapminder[gapminder$country == &#39;India&#39;,]
india_subset</code></pre>
<pre><code>    country year        pop continent lifeExp gdpPercap
697   India 1952  372000000      Asia  37.373  546.5657
698   India 1957  409000000      Asia  40.249  590.0620
699   India 1962  454000000      Asia  43.605  658.3472
700   India 1967  506000000      Asia  47.193  700.7706
701   India 1972  567000000      Asia  50.651  724.0325
702   India 1977  634000000      Asia  54.208  813.3373
703   India 1982  708000000      Asia  56.596  855.7235
704   India 1987  788000000      Asia  58.553  976.5127
705   India 1992  872000000      Asia  60.223 1164.4068
706   India 1997  959000000      Asia  61.765 1458.8174
707   India 2002 1034172547      Asia  62.879 1746.7695
708   India 2007 1110396331      Asia  64.698 2452.2104</code></pre>
<pre class="r"><code># check the dimension of the data frame
dim(india_subset)</code></pre>
<pre><code>[1] 12  6</code></pre>
<pre class="r"><code># get a summary of the data frame
summary(india_subset)</code></pre>
<pre><code>   country               year           pop             continent            lifeExp        gdpPercap     
 Length:12          Min.   :1952   Min.   :3.720e+08   Length:12          Min.   :37.37   Min.   : 546.6  
 Class :character   1st Qu.:1966   1st Qu.:4.930e+08   Class :character   1st Qu.:46.30   1st Qu.: 690.2  
 Mode  :character   Median :1980   Median :6.710e+08   Mode  :character   Median :55.40   Median : 834.5  
                    Mean   :1980   Mean   :7.011e+08                      Mean   :53.17   Mean   :1057.3  
                    3rd Qu.:1993   3rd Qu.:8.938e+08                      3rd Qu.:60.61   3rd Qu.:1238.0  
                    Max.   :2007   Max.   :1.110e+09                      Max.   :64.70   Max.   :2452.2  </code></pre>
</div>
<div id="reordering" class="section level3">
<h3>Reordering</h3>
<p>With vectors we saw the <code>sort()</code> function returned the
reordered vector, and that the <code>order()</code> function returned
the indices giving the order of the reordered vector. Let’s use the
<code>order()</code> function to change the ordering of
<code>india_subset</code> so that the most recent years are at the top
of the table.</p>
<p>First look at the help for <code>?sort</code> and <code>?order</code>
and note the <code>decreasing</code> parameter. If we want the years to
be decreasing, we should set this parameter to <code>TRUE</code>.</p>
<pre class="r"><code>sort(india_subset$year, decreasing = TRUE)</code></pre>
<pre><code> [1] 2007 2002 1997 1992 1987 1982 1977 1972 1967 1962 1957 1952</code></pre>
<p>This is the right ordering we want, but but we want the indices so
that we can reorder the rows of <code>india_subset</code>. Using
<code>order()</code> to give the row indices in the correct order gets
us where we want to go:</p>
<pre class="r"><code>india_subset_decreasing = india_subset[order(india_subset$year, decreasing = TRUE), ]
india_subset_decreasing</code></pre>
<pre><code>    country year        pop continent lifeExp gdpPercap
708   India 2007 1110396331      Asia  64.698 2452.2104
707   India 2002 1034172547      Asia  62.879 1746.7695
706   India 1997  959000000      Asia  61.765 1458.8174
705   India 1992  872000000      Asia  60.223 1164.4068
704   India 1987  788000000      Asia  58.553  976.5127
703   India 1982  708000000      Asia  56.596  855.7235
702   India 1977  634000000      Asia  54.208  813.3373
701   India 1972  567000000      Asia  50.651  724.0325
700   India 1967  506000000      Asia  47.193  700.7706
699   India 1962  454000000      Asia  43.605  658.3472
698   India 1957  409000000      Asia  40.249  590.0620
697   India 1952  372000000      Asia  37.373  546.5657</code></pre>
</div>
</div>
<div id="coercing-data" class="section level2">
<h2>Coercing data</h2>
<p>When we looked at <code>summary(gapminder)</code> and
<code>str(gapminder)</code>, we noticed the type for continent was
<code>character</code>, but perhaps a <code>factor</code> is more
appropriate here because there are only a small number of possible
values. We can “coerce” the <code>continent</code> column of
<code>gapminder</code> to be a factor using the <code>factor()</code>
function.</p>
<pre class="r"><code># Coerce the continent column to a factor
gapminder$continent &lt;- factor(gapminder$continent)</code></pre>
<p>And let’s see how the result of <code>summary()</code> and
<code>str()</code> change in response:</p>
<pre class="r"><code>summary(gapminder)</code></pre>
<pre><code>   country               year           pop               continent      lifeExp        gdpPercap       
 Length:1704        Min.   :1952   Min.   :6.001e+04   Africa  :624   Min.   :23.60   Min.   :   241.2  
 Class :character   1st Qu.:1966   1st Qu.:2.794e+06   Americas:300   1st Qu.:48.20   1st Qu.:  1202.1  
 Mode  :character   Median :1980   Median :7.024e+06   Asia    :396   Median :60.71   Median :  3531.8  
                    Mean   :1980   Mean   :2.960e+07   Europe  :360   Mean   :59.47   Mean   :  7215.3  
                    3rd Qu.:1993   3rd Qu.:1.959e+07   Oceania : 24   3rd Qu.:70.85   3rd Qu.:  9325.5  
                    Max.   :2007   Max.   :1.319e+09                  Max.   :82.60   Max.   :113523.1  </code></pre>
<pre class="r"><code>str(gapminder)</code></pre>
<pre><code>&#39;data.frame&#39;:   1704 obs. of  6 variables:
 $ country  : chr  &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; &quot;Afghanistan&quot; ...
 $ year     : int  1952 1957 1962 1967 1972 1977 1982 1987 1992 1997 ...
 $ pop      : num  8425333 9240934 10267083 11537966 13079460 ...
 $ continent: Factor w/ 5 levels &quot;Africa&quot;,&quot;Americas&quot;,..: 3 3 3 3 3 3 3 3 3 3 ...
 $ lifeExp  : num  28.8 30.3 32 34 36.1 ...
 $ gdpPercap: num  779 821 853 836 740 ...</code></pre>
<p>Notice that we got some additionally useful information from
<code>summary()</code> from the coercion. Namely, we get the number of
entries in <code>gapminder</code> that are on each continent.</p>
<blockquote>
<h2 id="note-stringsasfactors">Note: StringsAsFactors</h2>
<p>There are explicit forms of coercion, as in our factor example above.
But there are also implicit coercions. The most famous example of this
in R is the <code>stringsAsFactors</code> parameter of
<code>read.table()</code>. Prior to R 4.0, when importing a data frame
the <code>stringsAsFactors</code> argument was <code>TRUE</code> by
default, which caused all character columns to be factors by
<strong>default</strong>. This wasn’t a good default behavior, and the
default is now <code>FALSE</code>.</p>
</blockquote>
</div>
<div id="saving-your-data-frame-to-a-file" class="section level2">
<h2>Saving your data frame to a file</h2>
<p>We can save data to a file. We will save our
<code>india_subset</code> object to a .csv file using the
<code>write.csv()</code> function:</p>
<pre class="r"><code>write.csv(india_subset, file = &quot;results/india_subset.csv&quot;)</code></pre>
<p>The <code>write.csv()</code> function has some additional arguments
listed in the help, but at a minimum you need to tell it what data frame
to write to file, and give a path to a file name in quotes (if you only
provide a file name, the file will be written in the current working
directory).</p>
</div>
</div>
<div id="data-tips" class="section level1">
<h1>Data tips</h1>
<p>At the beginning of the lesson we suggested three things when dealing
with data in R:</p>
<ol style="list-style-type: decimal">
<li>Keep raw data separate from analyzed data</li>
<li>Keep spreadsheet data Tidy</li>
<li>Trust but verify</li>
</ol>
<p>As we all know, data can be messy–errors can be made when entering
it–and at some point those errors need to be corrected. To elaborate on
“Trust but verify,” consider the fact that in the <code>gapminder</code>
data set we had many countries over many continents, and any misspelling
could cause problems when summarizing the data. For instance, “North
America” could accidentally be entered as “NorthAmerica” or “North
america”. When reading in data, it’s a good idea to check that there
aren’t any odd things happening like this.</p>
<p>For character / categorical data, simply looking at the unique
elements of a column can help detect errors quickly:</p>
<pre class="r"><code>unique(gapminder$continent)</code></pre>
<pre><code>[1] Asia     Europe   Africa   Americas Oceania 
Levels: Africa Americas Asia Europe Oceania</code></pre>
<p>In this case, we don’t see any obvious misspellings that could cause
problems. For numeric data, looking at a summary of the column to see
the min/max values, as well as the mean/median, can help determine if
any obvious errors are present.</p>
<pre class="r"><code>summary(gapminder$pop)</code></pre>
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
6.001e+04 2.794e+06 7.024e+06 2.960e+07 1.959e+07 1.319e+09 </code></pre>
<p>In this case, we don’t notice any values that might be out of the
ordinary. For example, we know that there are countries with over a
billion people, so seeing a maximum of 1,319,000,000 is not strange. At
the same time, we know there are very small countries, so seeing a
minimum of 60,000 seems reasonable too. What would be unreasonable would
be a negative population, for example.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
